### ISSUE TITLE
Nullable table name

### ISSUE THREAD

**Issue by webmake** [laravel/framework#22409]
- Laravel Version: 5.5.22
- PHP Version: 7.1
- Database Driver & Version: sqlite

### Description:
I suspect that https://github.com/laravel/framework/pull/22222/files
after this change, stopped working tests, due table cache. Because in 5.5.21 working fine

### Steps To Reproduce:
In test file:
```
        $user = factory(User::class)->create();
        $address = factory(Address::class)->create([
            'user_id' => $user->id,
        ]);

        $this->initPHPUnitMock(SomeService::class)
            ->expects($this->once())
            ->method('myMethod')
            ->with($user, $address);
       this->app->make(OtherService::class)->otherMethod($user->id);
```
in code:
```
        $user = $this->finderService->getUser($id);
        $address = $user ->address;
        $this->someService->myMethod($user, $address);
```
Relation in user:
```
    public function address(): HasOne
    {
        return $this->hasOne(Address::class, 'user_id', 'id')
            ->where('main', true);
    }
```
Gives an error:
```
Expectation failed for method name is equal to "myMethod" when invoked 1 time(s)
Parameter 2 for invocation App\Services\SomeService::myMethod(App\User Object (...), App\Address Object (...)): void does not match expected value.
Failed asserting that two objects are equal. <Click to see difference>
```
Diff:
```
App\Address Object (
    'casts' => Array (...)
    'fillable' => Array (...)
    'connection' => 'sqlite_testing'
    'table' => 'addresses'
    'primaryKey' => 'id'
    'keyType' => 'int'
    'incrementing' => true
    'with' => Array ()
    'withCount' => Array ()
    'perPage' => 15
    'exists' => true
    'wasRecentlyCreated' => false
    'attributes' => Array (...)
    'original' => Array (...)
    'changes' => Array ()
    'dates' => Array ()
    'dateFormat' => null
    'appends' => Array ()
    'dispatchesEvents' => Array ()
    'observables' => Array ()
    'relations' => Array ()
    'touches' => Array ()
    'timestamps' => true
    'hidden' => Array ()
    'visible' => Array ()
    'guarded' => Array (...)
)
vs
App\Address Object (
    'casts' => Array (...)
    'fillable' => Array (...)
    'connection' => 'sqlite_testing'
    'table' => null
    'primaryKey' => 'id'
    'keyType' => 'int'
    'incrementing' => true
    'with' => Array ()
    'withCount' => Array ()
    'perPage' => 15
    'exists' => true
    'wasRecentlyCreated' => false
    'attributes' => Array (...)
    'original' => Array (...)
    'changes' => Array ()
    'dates' => Array ()
    'dateFormat' => null
    'appends' => Array ()
    'dispatchesEvents' => Array ()
    'observables' => Array ()
    'relations' => Array ()
    'touches' => Array ()
    'timestamps' => true
    'hidden' => Array ()
    'visible' => Array ()
    'guarded' => Array (...)
)
```
As you can see now it returns null, instead of real table

**Comment by webmake** [comment]
Yes, it is. Mentioned commit caused that error.
Seems that, untill this change, table attribute was null, so comparison was `null === null`. So now `$address = factory(Address::class)->create()` object is with filled table name (no problem), but problem that fetching via relation does not work in same way as via `create` method.

I guess hydrate was missed in that commit, because here is creating new instance, without transfering all data from $this (builder) to $instance variable
```
    public function hydrate(array $items)
    {
        $instance = $this->newModelInstance();

        return $instance->newCollection(array_map(function ($item) use ($instance) {
            return $instance->newFromBuilder($item);
        }, $items));
    }
```
In other words `$this->query->from` is `addreses`, while `$instance->table` is still `null`, or even getTable should be called in `newFromBuilder` method, I don't know

**Comment by webmake** [comment]
Problem is, that calling `$address = $user->address;` via relation, that object doesn't have filled table, so objects are inequal

again, full POC:
Example.php
```
<?php
namespace Tests\Unit;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Foundation\Testing\DatabaseMigrations;
use Tests\CreatesApplication;

/**
 * Class AuthenticateAsTest
 * @package Tests\Unit
 */
class ExampleTest extends \Illuminate\Foundation\Testing\TestCase
{
    use DatabaseMigrations, CreatesApplication;

    /**
     * @throws \InvalidArgumentException
     * @throws \LogicException
     * @throws \PHPUnit\Framework\Exception
     * @throws \PHPUnit\Framework\MockObject\RuntimeException
     * @throws \ReflectionException
     */
    public function test_relations(): void
    {
        /** @var \Illuminate\Database\Eloquent\Factory $factory */
        app(\Illuminate\Database\Eloquent\Factory::class)->define(User::class, function(\Faker\Generator $faker) {
            return [
                'id' => $faker->unique()->randomNumber(5),
            ];
        })->define(Address::class, function(\Faker\Generator $faker) {
            return [
                'id' => $faker->unique()->randomNumber(5),
                'main' => $faker->boolean,
            ];
        });
        $user = factory(User::class)->create();
        $address = factory(Address::class)->create([
            'user_id' => $user->id,
            'main' => true,
        ]);
        $address->wasRecentlyCreated = false;

        $builder = $this->getMockBuilder(SomeService::class)->disableOriginalConstructor();
        $mock = $builder->getMock();

        $this->app->instance(SomeService::class, $mock);

        $mock->expects($this->once())
            ->method('myMethod')
            ->with($address);
        $this->app->make(OtherService::class)->otherMethod($user->id);
    }
}

class SomeService
{
    public function myMethod($address)
    {
    }
}

class OtherService
{
    public function otherMethod(int $id)
    {
        $user = app(User::class)->newQuery()->find($id);
        $address = $user->address;
        app(SomeService::class)->myMethod($address->fresh());
    }
}

class User extends Model
{
    public $timestamps = false;

    public function address(): HasOne
    {
        return $this->hasOne(Address::class, 'user_id', 'id')
            ->where('main', true);
    }
}

class Address extends Model
{
    public $timestamps = false;
}
```
2017_12_13_000000_create_test_relations.php
```
<?php
declare(strict_types = 1);

use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;

class CreateTestRelations extends Migration
{
    /**
     * Run the migrations.
     *
     * @return void
     */
    public function up(): void
    {
        Schema::create('users', function(Blueprint $table) {
            $table->increments('id');
        });
        Schema::create('addresses', function(Blueprint $table) {
            $table->increments('id');
            $table->unsignedInteger('user_id');
            $table->boolean('main');
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down(): void
    {
        // Not necessary
    }
}
```
Failure diff:
![image](https://user-images.githubusercontent.com/5016464/34128683-2fc84f00-e44a-11e7-9221-a734b014dbc6.png)

Since 5.5.22 that test fails, but if you install 5.5.21 - this test pass with green.
Failure reason is exactly due this 
```
-            return str_replace('\\', '', Str::snake(Str::plural(class_basename($this))));
```
change into 
```
+	        $this->setTable(str_replace(
 +		        '\\', '', Str::snake(Str::plural(class_basename($this)))
 +	        ));
```
Even more I modified in my vendor that file back into `return` instead setTable, same test works as in previous laravel versions. So in my oppinion, that change was incomplete, because freshly loaded object via relation does not fill `table` field (seen on the right side in picture)

**Comment by armandsar** [comment]
Seeing same issue after upgrade to 5.2.22 from 5.2.21.

Simple way to replicate:

`
 $user = factory(User::class)->create()->fresh();
 $user = $user->fresh();

 $this->assertEquals($user, $user->fresh());
`

Will fail with table being null.

**Comment by webmake** [comment]
Nice @armandsar , your version is more shorter to replicate issue, I tested your variant

**Comment by themsaid** [comment]
Change reverted
